### ISSUE TITLE
Issue 5b: Forensics — Embedding space analysis

### ISSUE THREAD

**Issue by SriMed** [SriMed/rag-forensics#14]
# Issue 5b: Forensics — Embedding space analysis

## Description
Analyze where the query lives relative to retrieved chunks in embedding space. All other forensics dimensions are behavioral — they analyze what the retriever returned and how the model responded. This dimension operates at the representation level: it asks whether the query and retrieved chunks are geometrically coherent in the embedding space.

Like issue 5, this module produces **continuous metrics only** — no verdict. The metrics flow into the recommendation layer (issue 9) alongside signals from other forensics dimensions.

**The insight:** A high similarity score doesn't guarantee geometric coherence. Two chunks can score similarly against a query while being far apart from each other in embedding space — which signals the retriever is pulling from different semantic neighborhoods and the answer is synthesizing across unrelated concepts. Score distribution (issue 5) tells you *how confident* the retriever was. Embedding geometry tells you *whether that confidence was geometrically justified*. These are different questions.

**What exists:** 2D embedding visualizations exist as researcher tools. Shapley-based attribution (arxiv 2507.04480) attempts document-level influence scoring. Neither surfaces geometric coherence as a continuous diagnostic metric feeding an actionable recommendation system.

## Metrics to compute

```python
class EmbeddingSpaceMetrics(BaseModel):
    centroid_distance: float     # Distance from query to centroid of retrieved chunks.
                                 # Low = query is central to the retrieved cluster.
                                 # High = query is at the periphery — retriever pulled content semantically distant from the question.

    chunk_spread: float          # Mean pairwise cosine distance between retrieved chunk embeddings.
                                 # Low = chunks are semantically coherent, all from same neighborhood.
                                 # High = chunks are scattered across different semantic regions.

    query_isolation: float       # Ratio: query-to-centroid distance / mean chunk-to-centroid distance.
                                 # < 1.0 = query is closer to the cluster center than chunks are to each other (good).
                                 # > 1.0 = query is more isolated than the chunks themselves (retrieval pulled content that doesn't center on the question).

    projection: list[EmbeddingPoint]  # 2D PCA coordinates for visualization. One point per chunk + one for query.
```

```python
class EmbeddingPoint(BaseModel):
    label: str       # "query" or chunk_id
    x: float         # PCA component 1
    y: float         # PCA component 2
    is_query: bool
```

**Why each metric:**
- `centroid_distance`: single most interpretable geometric signal — is the query in the middle of what was retrieved?
- `chunk_spread`: measures semantic coherence of the retrieved set independently of the query
- `query_isolation`: normalized ratio that combines the two — values above 1.0 are the interesting failure case
- `projection`: 2D visualization that makes the geometry legible in the diagnostic card

## How metrics feed into the recommendation layer

These metrics are returned as a flat object alongside `RetrievalDistributionMetrics` and the other forensics outputs. Issue 9's `match_rule()` will use them as additional signal. Specifically:

- High `chunk_spread` + high `tail_mass` (from issue 5) → strong signal for R01 (ambiguous retrieval, reduce top-k or add reranker)
- High `query_isolation` + high `score_gap` (from issue 5) → R03 (decisive but wrong content — embedding model may not suit the domain)
- High `chunk_spread` + low `faithfulness` → reinforces R05 (noisy context reaching generator)

Add `embedding_space: EmbeddingSpaceMetrics` to `AnalyzeResponse`.

## Implementation details

- Create `services/forensics/embedding_analysis.py`
- `analyze_embedding_space(query_embedding: np.ndarray, chunk_embeddings: list[np.ndarray], chunk_ids: list[str]) -> EmbeddingSpaceMetrics`
- Function accepts **pre-computed embeddings** — no model calls inside this function
- Install: `poetry add scikit-learn` (for PCA)

### Computing each metric

```python
import numpy as np
from sklearn.decomposition import PCA
from sklearn.metrics.pairwise import cosine_distances

def analyze_embedding_space(query_embedding, chunk_embeddings, chunk_ids):
    Q = np.array(query_embedding).reshape(1, -1)
    C = np.array(chunk_embeddings)  # shape: (n_chunks, embedding_dim)
    n = len(C)

    # centroid_distance
    centroid = C.mean(axis=0, keepdims=True)
    centroid_distance = float(cosine_distances(Q, centroid)[0][0])

    # chunk_spread (mean pairwise cosine distance among chunks)
    if n > 1:
        pairwise = cosine_distances(C, C)
        upper_triangle = pairwise[np.triu_indices(n, k=1)]
        chunk_spread = float(upper_triangle.mean())
    else:
        chunk_spread = 0.0

    # query_isolation
    chunk_to_centroid_dists = cosine_distances(C, centroid).flatten()
    mean_chunk_centroid_dist = float(chunk_to_centroid_dists.mean()) if n > 0 else 1.0
    query_isolation = centroid_distance / mean_chunk_centroid_dist if mean_chunk_centroid_dist > 0 else 1.0

    # 2D PCA projection
    all_embeddings = np.vstack([Q, C])  # query first, then chunks
    pca = PCA(n_components=2)
    projected = pca.fit_transform(all_embeddings)

    points = [EmbeddingPoint(label="query", x=float(projected[0, 0]), y=float(projected[0, 1]), is_query=True)]
    for i, chunk_id in enumerate(chunk_ids):
        points.append(EmbeddingPoint(
            label=chunk_id,
            x=float(projected[i + 1, 0]),
            y=float(projected[i + 1, 1]),
            is_query=False
        ))

    return EmbeddingSpaceMetrics(
        centroid_distance=centroid_distance,
        chunk_spread=chunk_spread,
        query_isolation=query_isolation,
        projection=points
    )
```

### Threading query embeddings through the pipeline

The query embedding is already computed during ChromaDB retrieval in issue 3. Ensure `retrieve()` in `services/retriever.py` returns the query embedding alongside the chunks:

```python
class RetrievalResult(BaseModel):
    chunks: list[RetrievedChunk]
    query_embedding: list[float]  # add this field
```

Update `retrieve()` to return `RetrievalResult` instead of `list[RetrievedChunk]`. Update the `/analyze` router accordingly.

## Acceptance criteria
- [ ] `analyze_embedding_space()` returns `EmbeddingSpaceMetrics`
- [ ] `centroid_distance` is a float >= 0
- [ ] `chunk_spread` is 0.0 for single chunk input, no crash
- [ ] `query_isolation` is a float >= 0
- [ ] `projection` contains exactly `len(chunk_ids) + 1` points
- [ ] Exactly one point has `is_query=True`
- [ ] PCA computed on query + chunks together, not separately
- [ ] Function makes no embedding model calls — accepts pre-computed embeddings
- [ ] `retrieve()` updated to return `RetrievalResult` including `query_embedding`
- [ ] `AnalyzeResponse` updated to include `embedding_space` field
- [ ] `scikit-learn` added via `poetry add scikit-learn`

## TDD approach

```
tests/test_forensics/test_embedding_analysis.py
```

Use `numpy` to generate synthetic embeddings — no real model calls needed.

Tests to write before implementation:

1. Query at centroid of chunks → low `centroid_distance`, `query_isolation` < 1.0
2. Query far from all chunks → high `centroid_distance`, `query_isolation` > 1.0
3. All chunks identical → `chunk_spread` ≈ 0.0
4. Diverse chunks → `chunk_spread` > 0
5. `projection` length equals `len(chunk_ids) + 1`
6. Exactly one point has `is_query=True`
7. Single chunk input → `chunk_spread=0.0`, no crash
8. `centroid_distance` >= 0 always
9. `query_isolation` >= 0 always
10. Function does not instantiate or call any embedding model (assert no model mock called)
